numpy/core/src/multiarray/buffer.c

```diff
@@ -445,49 +458,22 @@ static PyObject *_buffer_info_cache = NULL;
 static _buffer_info_t*
 _buffer_info_new(PyObject *obj)
 {
+    /*
+     * Note that the buffer info is cached as pyints making them appear like
+     * unreachable lost memory to valgrind.
+     */
     _buffer_info_t *info;
     _tmp_string_t fmt = {NULL, 0, 0};
     int k;
     PyArray_Descr *descr = NULL;
     int err = 0;
 
-    /*
-     * Note that the buffer info is cached as pyints making them appear like
-     * unreachable lost memory to valgrind.
-     */
-    info = malloc(sizeof(_buffer_info_t));
-    if (info == NULL) {
-        PyErr_NoMemory();
-        goto fail;
-    }
-
-    if (PyArray_IsScalar(obj, Datetime) || PyArray_IsScalar(obj, Timedelta)) {
-        /*
-         * Special case datetime64 scalars to remain backward compatible.
-         * This will change in a future version.
-         * Note arrays of datetime64 and structured arrays with datetime64
-         * fields will not hit this code path and are currently unsupported
-         * in _buffer_format_string.
-         */
-        if (_append_char(&fmt, 'B') < 0) {
-            goto fail;
-        }
-        if (_append_char(&fmt, '\0') < 0) {
-            goto fail;
-        }
-        info->ndim = 1;
-        info->shape = malloc(sizeof(Py_ssize_t) * 2);
-        if (info->shape == NULL) {
+    if (PyArray_IsScalar(obj, Void)) {
+        info = malloc(sizeof(_buffer_info_t));
+        if (info == NULL) {
             PyErr_NoMemory();
             goto fail;
         }
-        info->strides = info->shape + info->ndim;
-        info->shape[0] = 8;
-        info->strides[0] = 1;
-        info->format = fmt.s;
-        return info;
-    }
-    else if (PyArray_IsScalar(obj, Generic)) {
```

Comment:
Removing this branch is going to cause custom dtypes to hit the assert below I think.

Reply:
The previous cleanup, disabled buffer handling for custom scalars completely (it never actually did anything anyway).  The scalar would have to implement its own buffer-handling now.

Reply:
I have added a test that rational raises a type error.

Reply:
Mind linking to the code that raises that TypeError?

Reply:
Eric there is no code. We simply do not implement the buffer interface in `Generic` at all, we only implement it for the concrete scalar instances. So the buffer interface is not defined and the default error is given.

Reply:
This code can only be hit from `void_getbuffer` and the arrays buffer getting, and `void_getbuffer` is obviously only set for the `void` type specifically.

Reply:
I think I must be looking at out of date code on my local machine then, apologies.